builtins/amp-img.js

```diff
@@ -132,10 +145,66 @@ export class AmpImg extends BaseElement {
 
     this.propagateAttributes(ATTRIBUTES_TO_PROPAGATE, this.img_);
     guaranteeSrcForSrcsetUnsupportedBrowsers(this.img_);
+    if (isExperimentOn(this.getAmpDoc().win, 'amp-img-auto-sizes')) {
+      this.maybeGenerateSizes_();
+    }
     this.applyFillContent(this.img_, true);
+
     this.element.appendChild(this.img_);
   }
 
+  /**
+   * This function automatically generates sizes for amp-imgs without
+   * the sizes attribute.
+   * @private
+   */
+  maybeGenerateSizes_() {
+    if (!this.img_) {
+      return;
+    }
+    // No need to generate sizes if already present.
+    const sizes = this.element.getAttribute('sizes');
+    if (sizes && sizes !== 'auto') {
+      return;
+    }
+    // Sizes is useless without the srcset attribute or if the srcset
+    // attribute uses the x descriptor.
+    const srcset = this.element.getAttribute('srcset');
+    if (!srcset || RegExp('[0-9]+x,.*|[0-9]+x$').test(srcset)) {
+      return;
+    }
+
+    const width = this.getLayoutWidth();
+    const viewportWidth = this.getViewport().getWidth();
+
+    const entry = `(max-width: ${viewportWidth}px) ${width}px, `;
+    const ratio = Math.round(width * 100 / viewportWidth);
+
+    let defaultSize = width + 'px';
+
+    if (this.getLayout() !== Layout.FIXED) {
```

Comment:
I think if the image is fixed size, it should probably not have a srcset--in which case the previous conditions should not match and we should never get into the situation of trying to define a `sizes` attribute anyway. But in the case that it does, `sizes` serves as a hint to the browser to decide which image from the `srcset` attribute to fetch.

Reply:
This is actually great for `thumbnail -> lightbox` use-case.

`<amp-img lightbox layout=fixed width=200 height=200 srcset="thumb-res.jpg 200px,  high-res.jpg: 600px ">`

now picks `thumb-res.jpg` for inline thumbnail instead of `high-res.jpg`  and lightbox upgrades it to `high-res.jpg`

Reply:
How? `layoutWidth` shouldn't change if it's fixed, and `viewportWidth` is constant.

Reply:
Right. I forgot about that. Yes, excellent point.

Reply:
`amp-image-lightbox` picks a different `src` based on `srcset` (not related to this change). This change forces `thumb-res.jpg` to be picked instead of `high-res.jpg` for thumbnail. I am missing something?

Reply:
I don't think lightbox has anything to do fixed-position elements getting a sizes attribute?

Reply:
It's not about lightbox here, without `sizes` on `fixed size` elements, browser would still pick the `src` based on `viewport` size, so for an image like:
```
<amp-img lightbox width=200 height=200 srcset="thumb-res.jpg 200px, high-res.jpg: 600px ">
```
browser would pick `high-res.jpg`, with this change adding proper `sizes`, browser would pick `thumb-res.jpg` .

Reply:
I don't see what is unexpected about that result.

Reply:
That's the whole point of introducing auto sizes, to pick the most appropriate image. Let's meet offline if you like the full background on this PR. Cathy and I have spend plenty of time discussing this and would be happy to share if you need the full details for the code review.

Reply:
Just to clarify this for posterity, the reason why its useful to have a `sizes` attribute whenever `srcset` is present is to prevent FF and Safari from automatically picking the highest resolution src, which is what they currently do per empirical experimentation. 

Why would we ever want to provide a `srcset` for a fixed layout image? The use case is lightboxing a small image. It's useful to provide a `srcset` for a fixed layout image so in case this image is lightboxed, `<amp-lightbox-gallery>` still has higher resolution images to upgrade to. 

Currently `<amp-lightbox-gallery>` has custom logic via `<amp-image-viewer>` to parse the `srcset` attribute from `<amp-img>` and manually set the `src` on a `<img>` tag based on the zoom factor. I was hoping to deprecate `<amp-image-viewer>` in favor of `<amp-pan-zoom>`, so this comment thread does raise a very good point that this proposed refactor might not necessarily work out of the box in context of FIXED layout images. I will note that in https://github.com/ampproject/amphtml/issues/20978